go-controller/pkg/node/controllers/egressip/egressip.go

```diff
@@ -0,0 +1,721 @@
+		nodeLister:         corelisters.NewNodeLister(nodeInformer.GetIndexer()),
+		nodeSynced:         nodeInformer.HasSynced,
+		namespaceLister:    namespaceInformer.Lister(),
+		namespaceSynced:    namespaceInformer.Informer().HasSynced,
+		podLister:          podInformer.Lister(),
+		podSynced:          podInformer.Informer().HasSynced,
+		routeManager:       routeManager,
+		linkManager:        linkmanager.NewController(nodeName, v4, v6),
+		ruleManager:        rulemanager.NewController(v4, v6),
+		iptablesManager:    iptables.NewController(v4, v6),
+		triggerReconcileCh: make(chan struct{}, 1),
+		nodeName:           nodeName,
+		v4:                 v4,
+		v6:                 v6,
+	}
+
+	for _, resourceInformer := range []struct {
+		resourceName string
+		cache.SharedIndexInformer
+	}{
+		{
+			"egressip",
+			eIPInformer.Informer(),
+		},
+		{
+			"node",
+			nodeInformer,
+		},
+		{
+			"namespace",
+			namespaceInformer.Informer(),
+		},
+		{
+			"pod",
+			podInformer.Informer(),
+		},
+	} {
+		_, err = resourceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.onAdd,
+			UpdateFunc: c.onUpdate,
+			DeleteFunc: c.onDelete,
+		})
+		if err != nil {
+			return c, fmt.Errorf("failed to add event hander for resource %q: %v", resourceInformer.resourceName, err)
+		}
+	}
+	return c, nil
+}
+
+func (c *Controller) Run(stopCh <-chan struct{}, wg *sync.WaitGroup) error {
+	klog.Info("Starting Egress IP Controller")
+
+	for _, resource := range []struct {
+		name string
+		cache.InformerSynced
+	}{
+		{
+			"egressip",
+			c.egressIPSynced,
+		},
+		{
+			"node",
+			c.nodeSynced,
+		},
+		{
+			"namespace",
+			c.namespaceSynced,
+		},
+		{
+			"pod",
+			c.podSynced,
+		},
+	} {
+		if !cache.WaitForNamedCacheSync(resource.name, stopCh, resource.InformerSynced) {
+			return fmt.Errorf("timed out waiting for %q caches to sync", resource.name)
+		}
+	}
+	// tell rule manager that we want to fully own all rules at a particular priority. Any rules created with this priority
+	// ,and we do not recognize it, will be removed.
+	if err := c.ruleManager.OwnPriority(ipRulePriority); err != nil {
+		klog.Errorf("Egress IP controller: failed to own priority %d - stale entries maybe present: %v", ipRulePriority, err)
+	}
+
+	go func() {
+		var err error
+		for {
+			select {
+			case <-stopCh:
+				wg.Done()
+				return
+			case <-c.triggerReconcileCh:
+				if err = c.reconcile(); err != nil {
+					klog.Errorf("Failed to reconcile egress IP: %v", err)
+				}
+			}
+		}
+	}()
+
+	wg.Add(3)
+	go func() {
+		c.linkManager.Run(stopCh, 2*time.Minute)
+		wg.Done()
+	}()
+
+	go func() {
+		c.iptablesManager.Run(stopCh, 6*time.Minute)
+		wg.Done()
+	}()
+
+	go func() {
+		c.ruleManager.Run(stopCh, 5*time.Minute)
+		wg.Done()
+	}()
+
+	return nil
+}
+
+func (c *Controller) triggerReconcile() {
+	select {
+	case c.triggerReconcileCh <- struct{}{}:
+	default:
+	}
+}
+
+func (c *Controller) onAdd(_ interface{}) {
```

Comment:
yep - i didnt take the time to filter this further. I dont believe there will be a lot of egress IP events. I would like to focus on the pods/namespace especially. Do you want me to filter in this PR?

Reply:
Ill also filter for egressable nodes for sure. Thanks.

Reply:
In the current PR, I added filtering that I think should be sufficient for now. LMK if you agree.
I was also thinking the node eip controller should be disabled if we are in the cloud because we only support BM now... @trozet WDYT?

Reply:
Disabling this on cloud env should help. 
Also, in the current implementation, a node should only really soak up loads of CPU if its egressable. It will keep constantly checking though.
Idea: Improvement for the feature is when node not labeled with egress label, only listen to node updates insteads of from pods,namespaces etc. Should help improve this further.

Reply:
Can I reduce the amount of signals to one (when the current nodes obj changes to check if egress label applied) when node isnt egressable later?

Reply:
Reworked this to follow the exact same pattern that APB route took. PTAL.